File: frontend/src/types/calendar.ts
export interface CalendarEvent {
  id?: number;
  date: string;
  title: string;
  category?: string;
  
  // 时间段
  morning_7_8?: string;
  morning_8_9?: string;
  morning_9_10?: string;
  morning_10_11?: string;
  morning_11_12?: string;
  
  afternoon_12_13?: string;
  afternoon_13_14?: string;
  afternoon_14_15?: string;
  afternoon_15_16?: string;
  afternoon_16_17?: string;
  afternoon_17_18?: string;
  
  evening_18_19?: string;
  evening_19_20?: string;
  evening_20_21?: string;
  evening_21_22?: string;
  evening_22_23?: string;
  evening_23_24?: string;
  
  // 完成状态
  morning_completed?: boolean;
  afternoon_completed?: boolean;
  evening_completed?: boolean;
  
  productivity_score?: number;
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

export interface ViewMode {
  type: 'day' | 'week' | 'month' | 'year';
  date: Date;
}

export interface Theme {
  mode: 'geek' | 'minimal';
}

export interface TimeSlot {
  key: string;
  label: string;
  period: 'morning' | 'afternoon' | 'evening';
}

export const TIME_SLOTS: TimeSlot[] = [
  { key: 'morning_7_8', label: '7:00-8:00', period: 'morning' },
  { key: 'morning_8_9', label: '8:00-9:00', period: 'morning' },
  { key: 'morning_9_10', label: '9:00-10:00', period: 'morning' },
  { key: 'morning_10_11', label: '10:00-11:00', period: 'morning' },
  { key: 'morning_11_12', label: '11:00-12:00', period: 'morning' },
  
  { key: 'afternoon_12_13', label: '12:00-13:00', period: 'afternoon' },
  { key: 'afternoon_13_14', label: '13:00-14:00', period: 'afternoon' },
  { key: 'afternoon_14_15', label: '14:00-15:00', period: 'afternoon' },
  { key: 'afternoon_15_16', label: '15:00-16:00', period: 'afternoon' },
  { key: 'afternoon_16_17', label: '16:00-17:00', period: 'afternoon' },
  { key: 'afternoon_17_18', label: '17:00-18:00', period: 'afternoon' },
  
  { key: 'evening_18_19', label: '18:00-19:00', period: 'evening' },
  { key: 'evening_19_20', label: '19:00-20:00', period: 'evening' },
  { key: 'evening_20_21', label: '20:00-21:00', period: 'evening' },
  { key: 'evening_21_22', label: '21:00-22:00', period: 'evening' },
  { key: 'evening_22_23', label: '22:00-23:00', period: 'evening' },
  { key: 'evening_23_24', label: '23:00-24:00', period: 'evening' }
];

File: backend/app/api/calendar.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime, date, timedelta
from pydantic import BaseModel

from ..core.database import get_db
from ..models.calendar import CalendarEvent

router = APIRouter()


# Pydantic 模型
class CalendarEventCreate(BaseModel):
    date: datetime
    title: str
    category: Optional[str] = None
    morning_7_8: Optional[str] = None
    morning_8_9: Optional[str] = None
    morning_9_10: Optional[str] = None
    morning_10_11: Optional[str] = None
    morning_11_12: Optional[str] = None
    afternoon_12_13: Optional[str] = None
    afternoon_13_14: Optional[str] = None
    afternoon_14_15: Optional[str] = None
    afternoon_15_16: Optional[str] = None
    afternoon_16_17: Optional[str] = None
    afternoon_17_18: Optional[str] = None
    evening_18_19: Optional[str] = None
    evening_19_20: Optional[str] = None
    evening_20_21: Optional[str] = None
    evening_21_22: Optional[str] = None
    evening_22_23: Optional[str] = None
    evening_23_24: Optional[str] = None
    notes: Optional[str] = None


class CalendarEventResponse(BaseModel):
    id: int
    date: datetime
    title: str
    category: Optional[str]
    morning_7_8: Optional[str]
    morning_8_9: Optional[str]
    morning_9_10: Optional[str]
    morning_10_11: Optional[str]
    morning_11_12: Optional[str]
    afternoon_12_13: Optional[str]
    afternoon_13_14: Optional[str]
    afternoon_14_15: Optional[str]
    afternoon_15_16: Optional[str]
    afternoon_16_17: Optional[str]
    afternoon_17_18: Optional[str]
    evening_18_19: Optional[str]
    evening_19_20: Optional[str]
    evening_20_21: Optional[str]
    evening_21_22: Optional[str]
    evening_22_23: Optional[str]
    evening_23_24: Optional[str]
    morning_completed: bool
    afternoon_completed: bool
    evening_completed: bool
    productivity_score: float
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


@router.get("/events", response_model=List[CalendarEventResponse])
def get_events(
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    category: Optional[str] = None,
    db: Session = Depends(get_db),
):
    query = db.query(CalendarEvent)

    if start_date:
        query = query.filter(CalendarEvent.date >= start_date)
    if end_date:
        query = query.filter(CalendarEvent.date <= end_date)
    if category:
        query = query.filter(CalendarEvent.category == category)

    events = query.order_by(CalendarEvent.date).all()
    return events


@router.post("/events", response_model=CalendarEventResponse)
def create_event(event: CalendarEventCreate, db: Session = Depends(get_db)):
    db_event = CalendarEvent(**event.dict())
    db.add(db_event)
    db.commit()
    db.refresh(db_event)
    return db_event


@router.get("/events/{event_id}", response_model=CalendarEventResponse)
def get_event(event_id: int, db: Session = Depends(get_db)):
    event = db.query(CalendarEvent).filter(CalendarEvent.id == event_id).first()
    if event is None:
        raise HTTPException(status_code=404, detail="Event not found")
    return event


@router.put("/events/{event_id}", response_model=CalendarEventResponse)
def update_event(
    event_id: int, event: CalendarEventCreate, db: Session = Depends(get_db)
):
    db_event = db.query(CalendarEvent).filter(CalendarEvent.id == event_id).first()
    if db_event is None:
        raise HTTPException(status_code=404, detail="Event not found")

    for field, value in event.dict().items():
        setattr(db_event, field, value)

    db_event.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(db_event)
    return db_event


@router.delete("/events/{event_id}")
def delete_event(event_id: int, db: Session = Depends(get_db)):
    db_event = db.query(CalendarEvent).filter(CalendarEvent.id == event_id).first()
    if db_event is None:
        raise HTTPException(status_code=404, detail="Event not found")

    db.delete(db_event)
    db.commit()
    return {"message": "Event deleted successfully"}


# 获取周视图数据
@router.get("/week/{year}/{week}")
def get_week_view(year: int, week: int, db: Session = Depends(get_db)):
    # 计算周的开始和结束日期
    start_date = datetime.strptime(f"{year}-W{week:02d}-1", "%Y-W%W-%w").date()
    end_date = start_date + timedelta(days=6)

    events = (
        db.query(CalendarEvent)
        .filter(CalendarEvent.date >= start_date, CalendarEvent.date <= end_date)
        .order_by(CalendarEvent.date)
        .all()
    )

    return {"start_date": start_date, "end_date": end_date, "events": events}


# 获取统计数据
@router.get("/stats/{year}/{month}")
def get_monthly_stats(year: int, month: int, db: Session = Depends(get_db)):
    start_date = datetime(year, month, 1)
    if month == 12:
        end_date = datetime(year + 1, 1, 1)
    else:
        end_date = datetime(year, month + 1, 1)

    events = (
        db.query(CalendarEvent)
        .filter(CalendarEvent.date >= start_date, CalendarEvent.date < end_date)
        .all()
    )

    total_events = len(events)
    completed_morning = sum(1 for event in events if event.morning_completed)
    completed_afternoon = sum(1 for event in events if event.afternoon_completed)
    completed_evening = sum(1 for event in events if event.evening_completed)

    avg_productivity = (
        sum(event.productivity_score for event in events) / total_events
        if total_events > 0
        else 0
    )

    return {
        "total_events": total_events,
        "completed_sessions": {
            "morning": completed_morning,
            "afternoon": completed_afternoon,
            "evening": completed_evening,
        },
        "average_productivity": avg_productivity,
    }
